src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp

```diff
@@ -0,0 +1,212 @@
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
+
+  using argument_tags = tmpl::list<::Tags::DataBox>;
+
+  template <typename DbTagsList, typename Metavariables>
+  std::array<Flag, Dim> operator()(const db::DataBox<DbTagsList>& box,
+                                   Parallel::GlobalCache<Metavariables>& cache,
+                                   const ElementId<Dim>& element_id) const;
+
+  void pup(PUP::er& p) override;
+
+ private:
+  std::optional<double> target_abs_truncation_error_{};
+  std::optional<double> target_rel_truncation_error_{};
+};
+
+// Out-of-line definitions
+/// \cond
+
+template <size_t Dim, typename TensorTags>
+TruncationError<Dim, TensorTags>::TruncationError(
+    const std::optional<double> target_abs_truncation_error,
+    const std::optional<double> target_rel_truncation_error,
+    const Options::Context& context)
+    : target_abs_truncation_error_(target_abs_truncation_error),
+      target_rel_truncation_error_(target_rel_truncation_error) {
+  if (not target_abs_truncation_error.has_value() and
+      not target_rel_truncation_error.has_value()) {
+    PARSE_ERROR(context,
+                "Must specify AbsoluteTarget, RelativeTarget, or both");
+  }
+}
+
+template <size_t Dim, typename TensorTags>
+TruncationError<Dim, TensorTags>::TruncationError(CkMigrateMessage* msg)
+    : Criterion(msg) {}
+
+template <size_t Dim, typename TensorTags>
+template <typename DbTagsList, typename Metavariables>
+std::array<Flag, Dim> TruncationError<Dim, TensorTags>::operator()(
+    const db::DataBox<DbTagsList>& box,
+    Parallel::GlobalCache<Metavariables>& /*cache*/,
+    const ElementId<Dim>& /*element_id*/) const {
+  auto result = make_array<Dim>(Flag::Undefined);
+  const auto& mesh = db::get<domain::Tags::Mesh<Dim>>(box);
+  std::array<DataVector, Dim> power_monitors{};
+  // Check all tensors and all tensor components in turn. We take the
+  // highest-priority refinement flag in each dimension, so if any tensor
+  // component has a truncation error above the target, the element will
+  // increase p refinement in that dimension. And only if all tensor components
+  // still satisfy the target with the highest mode removed will the element
+  // decrease p refinement in that dimension.
+  tmpl::for_each<TensorTags>([&result, &box, &mesh, &power_monitors,
+                              target_abs_truncation_error =
+                                  target_abs_truncation_error_,
+                              target_rel_truncation_error =
+                                  target_rel_truncation_error_](
+                                 const auto tag_v) {
+    // Stop if we have already decided to refine every dimension
+    if (result == make_array<Dim>(Flag::IncreaseResolution)) {
+      return;
+    }
+    using tag = tmpl::type_from<std::decay_t<decltype(tag_v)>>;
+    const auto& tensor = db::get<tag>(box);
+    for (const DataVector& tensor_component : tensor) {
```

Comment:
Moved to the cpp. Left the loop over components for now (not trivial because `Tensor::const_iterator` is templated and not sure if we need norms over components at all. Can be done later if needed.)